### PR TITLE
ENH: Styler column and row styles

### LINKED ISSUE #35605: ENH: Styler for specific column styling
Pandas Styler already contains the functionality needed for specific `row` and `column` styling via CSS class referencing.

Firstly it can add `table_styles` within the preliminary `<style>` tag, based on given `selectors`.
Secondly it adds specific CSS classes to each element, be it, `heading0` `col1` `row0` etc.

The current solution to add styling to a specific column can be done in two ways:

1) **Using Apply**: this is kind of hinted at in the docs and how most people attack the problem initially.

```
df = pd.DataFrame(data=[[0,1],[1,2],[3,4]], columns=['A', 'B'])
s = df.style.set_uuid('_')
s.apply(lambda x: ['color: red' for v in x], subset=['A'])
s
```

The problem with this approach is that it generates a lot of HTML code, each cell needs an id, and each id is added in the
initial <style> tag as the means of identifying elements. Without manually specifying the UUID (which are typically about 45bytes long) this creates a lot of useless data transmission and slows down websites for large tables:

`'<style  type="text/css" >\n#T__row0_col0,#T__row1_col0,#T__row2_col0{\n            color:  red;\n            color:  red;\n            color:  red;\n        }</style><table id="T__" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >A</th>        <th class="col_heading level0 col1" >B</th>    </tr></thead><tbody>\n                <tr>\n                        <th id="T__level0_row0" class="row_heading level0 row0" >0</th>\n                        <td id="T__row0_col0" class="data row0 col0" >0</td>\n                        <td id="T__row0_col1" class="data row0 col1" >1</td>\n            </tr>\n            <tr>\n                        <th id="T__level0_row1" class="row_heading level0 row1" >1</th>\n                        <td id="T__row1_col0" class="data row1 col0" >1</td>\n                        <td id="T__row1_col1" class="data row1 col1" >2</td>\n            </tr>\n            <tr>\n                        <th id="T__level0_row2" class="row_heading level0 row2" >2</th>\n                        <td id="T__row2_col0" class="data row2 col0" >3</td>\n                        <td id="T__row2_col1" class="data row2 col1" >4</td>\n            </tr>\n    </tbody></table>'`

2) **Using Table Styles**: since `col0` is a class identifier added to each cell relevant to column 'A' one can use table styles to apply directly:

```
df = pd.DataFrame(data=[[0,1],[1,2],[3,4]], columns=['A', 'B'])
s = Styler(df, uuid='_', cell_ids=False)
s.set_table_styles([{'selector': '.col0', 'props': [('color', 'red')]}])
s
```

This might be preferred in some instances since it can reduce the amount of HTML, particularly when combined with `cell_ids` argument. It only outputs the class type and can ignore individual ids, including their UUIDs.

`'<style  type="text/css" >
    #T__ .col0 {
          color: red;
    }</style><table id="T__" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >A</th>        <th class="col_heading level0 col1" >B</th>    </tr></thead><tbody>
                <tr>
                        <th id="T__level0_row0" class="row_heading level0 row0" >0</th>
                        <td  class="data row0 col0" >0</td>
                        <td  class="data row0 col1" >1</td>
            </tr>
            <tr>
                        <th id="T__level0_row1" class="row_heading level0 row1" >1</th>
                        <td  class="data row1 col0" >1</td>
                        <td  class="data row1 col1" >2</td>
            </tr>
            <tr>
                        <th id="T__level0_row2" class="row_heading level0 row2" >2</th>
                        <td  class="data row2 col0" >3</td>
                        <td  class="data row2 col1" >4</td>
            </tr>
    </tbody></table>'`

Therefore the **only missing code** is to put this together in a function which automatically maps the column name to the added CSS class and appends it to the existing table styles. 

#### API breaking implications

This only provides new functionality and will not impact previous versions.

### PR DESCRIPTION
**Pull request by attack68**
- [x] closes #35605 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
i am not sure we want to be adding api like this, rather can you use a keyword arg to .set_table_styles?

**Comment by attack68** [comment]
> i am not sure we want to be adding api like this, rather can you use a keyword arg to .set_table_styles?

One of the problems with set_table_styles is that it is a one-time function. Any repeated activity will simply overwrite previous styling, so the user is required to custom build up all style functions and then add at the end. No column argument will avoid this complexity.

One of the functions in this commit is literally just an extension, `extend_table_styles` which doesn't overwrite. I think this will be necessary then even if an argument solution to the problem is sought or not.

I could modify the code to accept an argument `column` into `extend_table_styles`, which repeats the code from the secondary function:

```
def extend_table_styles(table_styles, column=None):
    if column is not None:
        _styles = []
        for s in table_styles:
            c = str(self.data.columns.get_loc(col))
            _styles.append(
                    {"selector": s["selector"] + ".col" + c, "props": s["props"]}
            )
    else:
        _styles = table_styles
    if self.table_styles is None:
            return self.set_table_styles(table_styles)
    self.table_styles.extend(table_styles)
    return self
```

This would work for a single column and could be integrated into a user loop over different columns.

##Discussion

But since the purpose of Styler is to provide display capability, and it currently does not offer the ability to directly attach named CSS classes to cells, or rows or columns, I do not understand why you would shy away from providing the functionality to directly style a column in the most efficient manner. This can be extended to rows later.

It seems to be quite a basic functionality that has been requested in other instances that is easily crafted with current tools, and the direct purpose of styling columns is quite an easy user functionality to grasp???

**Comment by attack68** [comment]
I had a think on your comment @jreback. The addition of multiple arguments to the `set_table_styles` method allowed all of this functionality to be rolled up into the same function, which I think is what you are after, and in hindsight I think it is better also.

It also had the advantage of being able to address row styling at the same time as column styling.

Tests and examples have been amended.

File: pandas/io/formats/style.py [hunk 1]
```diff
@@ -909,34 +909,82 @@ def set_caption(self, caption: str) -> "Styler":
...
+
         Parameters
         ----------
-        table_styles : list
-            Each individual table_style should be a dictionary with
-            ``selector`` and ``props`` keys. ``selector`` should be a CSS
-            selector that the style will be applied to (automatically
-            prefixed by the table's UUID) and ``props`` should be a list of
-            tuples with ``(attribute, value)``.
+        table_styles : list or dict
```
Comment (jreback) [inline pandas/io/formats/style.py:923]: can you add a versionchanged tag here 1.2

and versionadded for new args

File: pandas/io/formats/style.py [hunk 2]
```diff
@@ -909,34 +909,82 @@ def set_caption(self, caption: str) -> "Styler":
...
-        >>> df = pd.DataFrame(np.random.randn(10, 4))
+        >>> df = pd.DataFrame(np.random.randn(10, 4),
+        ...                   columns=['A', 'B', 'C', 'D'])
         >>> df.style.set_table_styles(
         ...     [{'selector': 'tr:hover',
         ...       'props': [('background-color', 'yellow')]}]
         ... )
-        """
-        self.table_styles = table_styles
+        >>> df.style.set_table_styles({
```
Comment (jreback) [inline pandas/io/formats/style.py:957]: can you put a blank line between cases; comments on each of the cases are also useful (e.g. adding column ...)

File: pandas/io/formats/style.py [hunk 3]
```diff
@@ -909,34 +909,82 @@ def set_caption(self, caption: str) -> "Styler":
...
+        ...            'props': [('color', 'red')]}],
+        ...     'B': [{'selector': 'td',
+        ...            'props': [('color', 'blue')]}]
+        ... }, overwrite=False)
+        >>> df.style.set_table_styles({
+        ...     0: [{'selector': 'td:hover',
+        ...          'props': [('font-size', '25px')]}]
+        ... }, axis=1, overwrite=False)
+        """
+        if isinstance(table_styles, dict):
```
Comment (jreback) [inline pandas/io/formats/style.py:968]: use is_dict_like

File: pandas/io/formats/style.py [hunk 4]
```diff
@@ -909,34 +909,82 @@ def set_caption(self, caption: str) -> "Styler":
...
+            if axis == 0 or axis == "index":
```
Comment (jreback) [inline pandas/io/formats/style.py:969]: axis in [...]

File: pandas/io/formats/style.py [hunk 5]
```diff
@@ -909,34 +909,82 @@ def set_caption(self, caption: str) -> "Styler":
...
+                obj = self.data.columns
+                idf = ".col"
+            else:
+                obj = self.data.index
+                idf = ".row"
+
+            _styles = []
```
Comment (jreback) [inline pandas/io/formats/style.py:976]: can you write this as a list comprehension

File: pandas/io/formats/style.py [hunk 6]
```diff
@@ -909,34 +909,82 @@ def set_caption(self, caption: str) -> "Styler":
...
+            `axis` argument. These will be mapped to row or col CSS
+            selectors. MultiIndex values as dict keys should be
+            in their respective tuple form. The dict values should be
+            a list as specified in the form with CSS selectors and
+            props that will be applied to the specified row or column.
+        axis : {0 or 'index', 1 or 'columns', None}, default 0
+            Apply to each column (``axis=0`` or ``'index'``), to each row
+            (``axis=1`` or ``'columns'``). Only used if `table_styles` is
+            dict.
+        overwrite : boolean, default True
```
Comment (jreback) [inline pandas/io/formats/style.py:940]: hmm, i am not sure we have this style of keyword anywhere else. @TomAugspurger ?

File: pandas/io/formats/style.py [hunk 7]
```diff
@@ -909,34 +909,82 @@ def set_caption(self, caption: str) -> "Styler":
...
+            Styles are replaced if `True`, or extended if `False`. CSS
+            rules are preserved so most recent styles set will dominate
```
Comment (jreback) [inline pandas/io/formats/style.py:942]: i generally DO like the ability to chain things.

File: pandas/tests/io/formats/test_style.py [hunk 8]
```diff
@@ -1682,6 +1682,26 @@ def f(a, b, styler):
         result = styler.pipe((f, "styler"), a=1, b=2)
         assert result == (1, 2, styler)
 
+    def test_chaining_table_styles(self):
+        df = pd.DataFrame(data=[[0, 1], [1, 2]], columns=["A", "B"])
```
Comment (jreback) [inline pandas/tests/io/formats/test_style.py:1686]: can you add the issue number (or PR number as i don't think there is an issue) on tests

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/formats/style.py (hunk 6 above)
Comment (attack68) [reply pandas/io/formats/style.py:946]: The reason I added the argument was for backwards compatibility.
As in I know this function has previously been used like:
`styler.set_table_styles(styler.table_styles.extend(new_styles))`.
I think it makes sense to allow the function dual purpose.

Comment (attack68) [reply pandas/io/formats/style.py:946]: I found this keyword used in `DataFrame.update()` and `DataFrame.combine()`. The use is a bit more specific but not dissimilar.

**Comment by attack68** [comment]
conversations/changes resolved or comments added, all green, @jreback .. ping.

**Comment by TomAugspurger** [comment]
Seems good to me.

**Comment by attack68** [comment]
@jreback reviving this after a couple of weeks, all green, and changes requested implemented.

**Review by TomAugspurger — APPROVED** [review]
Restarted the failed build.

**Comment by attack68** [comment]
@jreback keeping this one alive...

**Comment by jreback** [comment]
can you merge master. I understand what `overwrite` means here, but we don't actually use this *anywhere* else in pandas.

I think it might be better to use `inplace` (though not a fan of this either), and prefer chaining to append.

cc @TomAugspurger

**Comment by attack68** [comment]
This is a strange one.

The problem with `inplace` is that Pandas typically have object methods which return new objects, and the underlying object is unaffected. If `inplace` is True then the underlying object is permanently affected and the return is `None`.

In `Styler`, this is not really how it was designed or works. I would say there are 6 possibilities for how this method might work. For `set_table_styles(..)`:

i) <s>The underlying `Styler` is **unaffected** and a new `Styler` is returned with **only** the argument attributes set.</s>
ii) The underlying `Styler` is **affected with only** the argument attributes set and a self `Styler` returned .
iii) <s>The underlying `Styler` is **affected with only** the argument attributes set and no return value.</s>
iv) <s>The underlying `Styler` is **unaffected** and a new `Styler` is returned with the **new attributes extended** to the underlying.</s>
v) The underlying `Styler` is **affected with new attributes extended** and a self `Styler` is returned.
vi) <s>The underlying `Styler` is **affected with new attributes extended** to the underlying and no return value.</s>

Currently, `Styler` has no methods that ever do i) or iv). All methods always seem to update itself `inplace`.<br>
Also `Styler` never returns None, so this would seem to rule out iii) and vi).<br>

`set_table_styles` actually does ii) which is a kind of `inplace=True` with a return, which seems quite unconventional for Pandas. That is why I don't think the `inplace` argument really works here. And to get the additional behaviour of v) which is what we want left the only other options of either: adding the `overwrite` argument (see here.. https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.combine.html), or creating another method: `extend_table_styles` or `add_table_styles` would be better??

FYI part of the design structure I think was to maintain a self contained object as a property: `df.style` which then developed the ability to be manipulated but always `inplace` by design.

**Comment by TomAugspurger** [comment]
Fixed the merge conflict in the whatsnew. I think this was OK last I looked.

**Comment by attack68** [comment]
I merged this into master.

Only outstanding issue is acceptance or not of the `override` keyword. 

I stand by this, rather than `inplace` as explained above.

File: pandas/io/formats/style.py [hunk 9]
```diff
@@ -987,34 +987,95 @@ def set_caption(self, caption: str) -> "Styler":
         self.caption = caption
         return self
 
-    def set_table_styles(self, table_styles) -> "Styler":
+    def set_table_styles(self, table_styles, axis=0, overwrite=True) -> "Styler":
```
Comment (jreback) [inline pandas/io/formats/style.py:990]: can you type table_styles

**Review by jreback — CHANGES_REQUESTED** [review]
ok small comment, pls add a whatsnew note in 1.2, merge master and ping on green.

File: pandas/io/formats/style.py (hunk 9 above)
Comment (attack68) [reply pandas/io/formats/style.py:994]: I tried adding the following to func def, but it got errors, I don't know my way around mypy and this seems slightly more complicated than normal with the multiple acceptable input types and programming logic:

    table_styles: Union[
            List[Dict[str, Union[str, List[Tuple[str, str]]]]],
            Dict[
                Union[str, Tuple[Any, ...]],
                List[Dict[str, Union[str, List[Tuple[str, str]]]]],
            ],
        ] = []

So I reverted the attempt.

However, all green, pinging @jreback

**Review by jreback — APPROVED** [review]
ok this is fine. pls followup with a typing PR as the table_styles is very confusing w/o it

**Comment by jreback** [comment]
thanks @attack68

**Pull request merged**